File: tests/fixtures/typescript/dual_pool.ts
/**
 * Fixture for D8 Dual-Pool Detection — TypeScript patterns.
 */

import { Pool } from 'pg';

// --- D8: Dual pool pattern ---

// Restricted pool (RLS-aware) for user-facing routes
const rlsPool = new Pool({ connectionString: process.env.DATABASE_URL_APP });

// Admin pool for workers/migrations
const adminPool = new Pool({ connectionString: process.env.DATABASE_URL_ADMIN });

// WRONG: API handler using admin pool — D8 should flag
export async function getUserProfile(req: Request, res: Response) {
  const { rows } = await adminPool.query(
    'SELECT * FROM users WHERE id = $1',
    [req.params.id]
  );
  return res.json(rows[0]);
}

// CORRECT: API handler using restricted pool
export async function getUserSessions(req: Request, res: Response) {
  const { rows } = await rlsPool.query(
    'SELECT * FROM sessions WHERE user_id = $1',
    [req.user.id]
  );
  return res.json(rows);
}

// OK: Worker using admin pool (expected)
export async function cleanupExpiredSessions() {
  await adminPool.query('DELETE FROM sessions WHERE expires_at < NOW()');
}

File: tests/fixtures/python/dual_pool.py
"""Fixture for D8 Dual-Pool Detection — Python patterns."""

import os
from sqlalchemy import create_engine

# --- D8: Dual pool pattern (perseworks rlspool) ---

# Restricted pool for user-facing code (should use this)
app_engine = create_engine(os.getenv("DATABASE_URL_APP"))

# Admin pool for workers/migrations (should NOT be used in handlers)
admin_engine = create_engine(os.getenv("DATABASE_URL_ADMIN"))


# WRONG: Route handler using admin pool — D8 should flag this
async def get_user_profile(request):
    """User-facing endpoint using admin pool."""
    async with admin_engine.connect() as conn:
        result = await conn.execute(
            "SELECT * FROM users WHERE id = :id", {"id": request.user_id}
        )
        return result.fetchone()


# CORRECT: Route handler using restricted pool
async def get_user_sessions(request):
    """User-facing endpoint using restricted pool."""
    async with app_engine.connect() as conn:
        result = await conn.execute(
            "SELECT * FROM sessions WHERE user_id = :uid",
            {"uid": request.user_id},
        )
        return result.fetchall()


# OK: Worker using admin pool (expected)
async def cleanup_expired_sessions():
    """Background worker — admin pool is acceptable."""
    async with admin_engine.connect() as conn:
        await conn.execute("DELETE FROM sessions WHERE expires_at < NOW()")
